Code for Issue #97

CowedOffACliff opened the pull request:
This is code adding the functionality in Issue #97. Issues with automatic merging are largely due to reformatting of files upstream after I started working on this combined with the tests that were added to test files I was required to modify. I've already made an attempt to merge the branches on my end so I'm hopeful that it spits out less confusing garbage for you. When I attempted it on my end there were 4 total files in conflict.

Changes: A new job type has been added which is triggered by two valid entities breeding together. Two commands have been added to operate the functionality: jobcenter job addBreedable <job> <breedable entity> <price> and jobcenter job removeBreedable <job> <breedable entity>.

Small changes have also been made to the tab completion function that adds entities to the tab list for the regular addMob/removeMob commands. I used a test entity.isAlive() to eliminate a lot of useless things from the list that really don't belong like arrow entities and minecarts.

As a result of these changes the compile tests in two test files needed to be changed. In one file the new command types needed to be added in to the function which changed the total list size being tested for. The integers corresponding to the list sizes were raised by 2 in both cases. One of the cases tests for the total number of items returned in the tab completion of addMob/removeMob for which the integer value being tested for has been lowered from 104 to 67 to meet the change in functionality.

In UltimateEconomy.java the handleJobInfo() command has been modified. The new job type has been added and I have also modified the listing for Blocks to be consistent with other job types. The en_US resource bundle has been modified to have text corresponding to these additions and changes. Blocks are now prefixed with "Mine:" and the new breeding job type is prefixed with "Breed:". As a user I felt that the colon after the name of each descriptor combined with the prefixes created a more professional and consistent look for the data being displayed.

A static ArrayList now exists with a predefined list of all of the vanilla entities that can currently be bred (16 in total) which is used to cull entities from tab completion of the breedable commands. A TODO is marked here to add Striders and Hoglins to this list when Minecraft is officially updated to 1.16. Currently it's just positioned at the top of the JobSystemValidationHandler because its primary use was for a validation command. The list is also referenced when adding entities to the tab list for breedable commands. This data could probably be moved to a more elegant location but for now it's functional.

Currently Missing:
Things currently missing from this are the addition of new unit tests corresponding to the newly added commands. The tests required will be nearly identical to those for addMob/removeMob commands but probably with a different integer value (16 in 1.15, 18 in 1.16) being tested for the tab completion list. Resource bundles have also not been modified for any language but my native English.

CowedOffACliff committed:
Functions for handling a job for breeding entities together minus operational commands.

CowedOffACliff committed:
Time to see I forgot.

CowedOffACliff committed:
Working on commands. Tab completion was missing. Will not pass unit tests for Maven install.

CowedOffACliff committed:
Modify existing tests for added commands.

Changed the existing tests to include the new commands and build. No new corresponding tests added yet.

CowedOffACliff committed:
Improve entity tab completion and breedable list handling.

Improve tab completion of listed entities to account for EntityType.isAlive() which clears non living entities from the tab list. Modify tests to account for the difference.

Breedable adding and tab completing now checks against a predefined list of valid entities that can breed in vanilla minecraft.

CowedOffACliff committed:
This is the commit that finally works.

In this commit I am finally actually able to build the jar, load the plugin, and add and remove "breedables" to a job and then breed two entities together to get money.

CowedOffACliff committed:
Modify /jobinfo, modify language bundle.

This commit modifies the /jobinfo command to correctly display the new information. It also modifies the command to make listed blocks appear more consistent in the list. A colon : was added at the end of each prefix, "Mine:, Fish:, Kill, Breed:, to provide a cleaner look for the command. en_US language bundle is modified, other languages are not yet changed.

CowedOffACliff committed:
Reformatting

Attempting to reformat to aid in upstream merge.

CowedOffACliff committed:
Merge remote-tracking branch 'upstream/dev' into dev

CowedOffACliff committed:
Fix after merge attempt.

After merge from upstream fixed the things I broke while trying to resolve conflicts.

CowedOffACliff commented:
Looks like the only remaining conflict is in the validation handler which should just be additional code with no other changes.

CowedOffACliff committed:
Update .gitignore

CowedOffACliff committed:
Fix build.

Remove source files duplicating com.mojang.authlib because why were they there to begin with? These files are provided by Spigot as compiled code. Mark spigot artifact as 'provided' instead of 'test' in pom to fix the build. Previously the plugin was failing to load with a NoSuchMethodException because of compiler confusion over these conflicting files.

CowedOffACliff committed:
Bugfix for issue #98

This is a small fix to the system that load player banks. Previously if for whatever reason a player name existed in PlayerFile.yml that had neither a legacy bank account nor a modern iban it would retrieve a null iban value and then attempt to throw a GeneralEconomyException passed with a null value thus causing the plugin to crash with an NPE instead of handling the exception. New code checks for a null iban state and creates a fresh iban account with 0 currency.

CowedOffACliff committed:
Merge branch 'dev' into dev

CowedOffACliff commented:
Fixed conflicts with new code functions.
Added code to fix issue #98.
Removed source files for com.mojang.authlib from the project because they're duplicates of the one provided by the Spigot artifact and they just confuse the build. Marked Spigot artifact as 'provided' instead of 'test' so that it builds properly.

Give 'er a whirl, chief.

CowedOffACliff committed:
Another bugfix.

Commit bugfix for NPE crash when loading player shops.

ItemStack stack = getShopInventory().getItem(slot).clone()

This method is called and iterates on every item slot in the player shop inventory. When a stack contains nothing it is technically null. Instantiating the local variable in this way causes the code to pass a null object to the .clone() function which crashes with an unchecked NPE instead of just instantiating the stack as null.

Fix is to check for itemstack != null and then clone the stack.

CowedOffACliff committed:
Merge branch 'dev' of https://github.com/CowedOffACliff/Ultimate_Economy into dev

CowedOffACliff commented:
Fixed another NullPointer crash in the plugin loading process.

CowedOffACliff commented:
Shop UI is totally unresponsive. Commands can still edit the contents of the shops but attempting to purchase/sell items or trigger any click event fails completely. For some reason the item stack being passed via the InventoryClickEvent is null even if you are clicking a valid item.

catch441 commented:
the actual version inside the dev branch has some commented out code due refactoring, so some functionality is not working right now as you saw.

catch441 commented on src/main/java/com/ue/jobsystem/commands/JobTabCompleter.java:
```diff
@@ -36,24 +37,35 @@
+		// during the "job" part. Catch the exception and return a blank list
+		// then it won't throw stack traces to console.
+		try {
+			switch (args[1]) {
```
nice catch

catch441 commented on src/main/java/com/ue/jobsystem/impl/JobSystemValidationHandler.java:
```diff
@@ -17,6 +19,22 @@
+			EntityType.LLAMA, EntityType.TURTLE, EntityType.PANDA));
+	// TODO EntityType.HOGLIN, EntityType.STRIDER needed for 1.16
+	
+	public static ArrayList<EntityType> getBreedableTypes()
```
is it really necessary to be public? Is this method used outside the class?

CowedOffACliff replied:
It's used by the tab completion function to cull the tab completion list down to only these specific available entries. That's why I made it public. It really should be in a more elegant place for this purpose I just haven't written code in a long time and I wasn't certain which 'elegant' way was the most practical.

catch441 commented on src/main/java/com/ue/player/impl/EconomyPlayerImpl.java:
```diff
@@ -445,12 +445,20 @@ private void loadBankAccount() {
 			String iban = config.getString(getName() + ".Iban");
+			// at this code step if the iban is null the player has no existing account, make a new one
+			// safely handles a state where a player name has no account information causing an NPE
+			if (iban == null) {
```
I thrust you that this resolves a problem. but just a litle question: The if part of this statement (we are here in the else part) checks, if the player has the old savefile format with the money directly saved inside the playerfile and converts it to the new iban format. Why is it possible, that the new loading is triggered with the old format?

CowedOffACliff replied:
I'm not actually certain what causes it but the reason this was an issue is because the PlayerFile I had generated from running 1.2.5f on my server contained usernames for players but those player usernames didn't have any actual data attributed to them. The entire "Player:" section of the file was missing for those users. This would cause the code to reach a point where it had already checked to see if the player had any account information by checking the top of the file for the player's username in the list. Because the name was present it would then assume the player had actual account information but would attempt to parse it and end up with non-existent null data. The simple additional null check effectively catches this edge case with corrupted data and handles it cleanly by just bringing that user up to date.

catch441 commented on src/main/java/com/ue/shopsystem/impl/PlayershopImpl.java:
```diff
@@ -197,7 +197,8 @@ public Inventory getStockpileInventory() {
 	private void updateItemInStockpile(int slot) throws GeneralEconomyException, ShopSystemException {
-		ItemStack stack = getShopInventory().getItem(slot).clone();
+		ItemStack stack = null;
+		if (getShopInventory().getItem(slot) != null) stack = getShopInventory().getItem(slot).clone();
```
- inline if statements should be avoided
- every slot has a item, you can add a null item to a inventory, but if you get the item, it is a item with the material AIR. So this should not be necessary to check for null (I know I'm doing it in the next if statement too, the code is old and my knowledge about the minecraft code growed since then)

CowedOffACliff replied:
The reason for this null check is because the code was crashing on a NullPointerException when it attempted to load inventories. This happened consistently at the line
stack = getShopInventory()getItem(slot).clone();
After spending a few hours working through the problem I came to the conclusion that by calling the .clone() function on Object.clone() in this way passing the null value to this function specifically would throw the NPE. The local object <ItemStack stack> is still used throughout all of the code with a default assignment of null. The only thing that has changed is now it checks whether or not the item in the inventory slot is null or not before attempting to clone it. If the inventory slot is null even though it defaults to Material.AIR it doesn't actually contain any data that needs to be cloned anyway.

catch441 commented on src/main/resources/language/MessagesBundle_en_US.properties:
```diff
@@ -113,10 +113,13 @@ home_info = Your homes: {0}
-jobinfo_killprice = Kill {0} {1} {2}
+jobinfo_killprice = Kill: {0} {1} {2}
+# {0} = breedable entity {1} = price {2} = currency
+jobinfo_breedingprice = Breed: {0} {1} {2}
```
Unfortunately if we add new messages, we have to translate them into all other languages too. If we don't, players with other languages just see "jobinfo_breedingprice" in the chat

CowedOffACliff replied:
In this specific case it may be possible for us to translate them at least partially. The specific messages needed are _extremely_ close to the existing text for addMob and removeMob. Without knowing the language it's impossible to now the specific context of how they translated "mob" but I think it's not a bad idea for me to go ahead and just copy/paste these two lines to supplement.

CowedOffACliff replied:
It occurs to me that the language message code could be rewritten such that if a label entry is missing from a language file it attempts to parse the text from the English file as a default to supplement the line before spitting out the ex. "!jobinfo_breedingprice" empty text. It would probably be more beneficial to end users in the event a translation is missing. Perhaps it should be in combination with the debug text so that people have some way of knowing which label is missing for their language.

English is a good choice as a language to fall back on because people in a lot of other cultures will learn at least basic English as a foreign language.

catch441 commented on src/main/resources/language/MessagesBundle_en_US.properties:
```diff
@@ -164,6 +167,9 @@ job_join = You have joined the job {0}.
 jobcenter_addMob = The entity {0} was added to the job.
 jobcenter_removeMob = The entity {0} was removed from the job.
+# {0} = breedable entity name
+jobcenter_addBreedable = The breedable entity {0} was added to the job.
```
same as above

catch441 commented on src/test/java/com/ue/jobsystem/commands/JobCommandExecutorTest.java:
```diff
@@ -625,7 +625,8 @@ public void jobCommandTest() {
 		executor.onCommand(player, null, "jobcenter", args);
 		String message = player.nextMessage();
-		assertEquals("/jobcenter job [create/delete/addItem/removeItem/addMob/removeMob/addFisher/removeFisher]",
+		assertEquals(
```
please add tests for the new commands

catch441 commented on src/test/java/com/ue/jobsystem/commands/JobTabCompleterTest.java:
```diff
@@ -16,6 +16,7 @@
 import com.ue.jobsystem.api.JobcenterController;
+import com.ue.player.api.EconomyPlayerController;
 import com.ue.ultimate_economy.UltimateEconomy;
 
```
missing tests for new comments

catch441 reviewed (changes requested):
the logfiles should not be pushed into the branch. My comments are mostly identical with your "Currently Missing:" part.
I'm working on the refactoring and the unit tests for a while now, so the unit tests are necessary. After I finished my refactoring and my unit tests I planned to build a development team. If you wish you can join  :D

CowedOffACliff committed:
Update .gitignore

Removing logs per request

CowedOffACliff committed:
Remove logs

Logs removed

CowedOffACliff commented:
> the actual version inside the dev branch has some commented out code due refactoring, so some functionality is not working right now as you saw.

After messing with it for a while I realized that was the missing code and uncommented it just to see how broken the changes had made it. The answer was "hornet's nest." Without knowing all of changes that one is a project for you to finish. XD

CowedOffACliff commented:
Will work on getting tests added equivalent to tests for similar functions.
Will copy/paste text from addMob and removeMob commands into the various language files so that they will at least be functional for users until real translations can be donated.
The added line for the "Mine:" text is a bit of a problem to stew on because language is complicated but after looking at translations of the official Minecraft wiki it's apparent that for most languages poaching a translation of "remove" or "destroy" will probably suffice.

I'm just a guy not a professional so my implementations of things will not always be the best way of doing it. If you ever have a question about something I will usually be able to give a specific explanation of the exact reason for a piece of code and what I encountered that made me determine a need for it.

As far as actually working on the project I have a lot to do and not a crazy amount of free time, but I am running a server with this plugin and I see its potential.  I've got some motivation for improving it and seeing if in time I can work out how to implement some of the more difficult things I'd like to see added or upgraded. Personally I'm aiming to get a stable feature complete version of the development branch that I can run in the live environment so that I can get features directly to the server for tinkering.

CowedOffACliff committed:
Fixing language files part 1

Went into all the existing language files and added text for all of the modifications I made. For some languages I have gone through with translations from various sources combined with modifications of pre-existing translation for related info and added new translated text. I intend to do as many of these as I reasonably can.

CowedOffACliff committed:
Add some tests

Commit a couple tests; motivate self to add the rest of them.

CowedOffACliff commented:
Basically waiting for the Paper build of 1.16 to come out before I get motivated to finish this and go through another merge.

catch441 commented:
merged
